# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/VenkataVishnuVardhanReddy/Flyrank/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

**Selected Lane:** Lane 2 — Refresh / Content Opportunity Scoring

**Why this lane?**
Organic search traffic decline is one of the most critical and direct threats to digital content publishers. Prioritizing which content assets to update, expand, or prune represents a high-leverage decision because editorial resources are strictly finite. We pick this lane because the starter data indicates a high volume of declining pages (over 54%), many of which retain substantial visibility. A learned prioritization model offers a significant efficiency lift over a basic manual refresh rule, allowing editors to capture traffic opportunities before they are lost to competing pages.

In [1]:
# Selected lane confirmation
print("Provisional Lane: Lane 2 — Refresh / Content Opportunity Scoring")


Provisional Lane: Lane 2 — Refresh / Content Opportunity Scoring


## 2. The question: decision, action, cost of a wrong call

- **What decision does this improve?**
  Deciding which specific pages (content assets) in a client's portfolio should be assigned to an editor for refresh review first.
- **Who acts on the output, and what do they do?**
  A content editor or content manager acts on the ranked list. They review the page's search console and analytics metrics alongside its reason codes (e.g., `thin_visible_page`, `page_one_decay_risk`) and determine the specific update type (e.g., rewrite title/meta, add content depth, or monitor it).
- **What does a wrong answer cost?**
  - *False Positive (recommending a healthy or stable page for refresh)*: Wasted writing/editorial hours (often valued at $100–$300 per article) on content that did not need updates.
  - *False Negative (missing a page that is in severe decay)*: Compounded loss of search ranking position, organic traffic, and conversion revenue over time.
- **Why does data or ML help at all?**
  A simple rule (e.g., "refresh every page older than 180 days") is inefficient because content decay is non-linear and multidimensional. Some old pages remain highly stable, while some newer pages decay rapidly due to search keyword volume shifts. ML allows us to weight and combine multiple signals (impressions, position drift, CTR, and scroll/engagement rates) to capture complex decay patterns that simple manual thresholds miss.

In [2]:
import pandas as pd
df = pd.read_csv("../../data/raw/content_refresh_anonymized.csv")
decline_rate = df["trend_direction"].str.lower().eq("down").mean()
print(f"Base rate of declining pages in the dataset: {decline_rate:.3%}")


Base rate of declining pages in the dataset: 54.207%


## 3. Quick look at the data (2-3 real numbers)

To confirm that Lane 2 is a high-leverage direction, we computed key metrics from the starter dataset:
1. **Decline Prevalence**: Out of 30,000 pages, **16,262 (54.2%)** are currently in a declining trend.
2. **High-Visibility Exposure**: **9,961** of these declining pages (61.3%) still maintain high search console visibility (`impressions_90d >= 500`). These represent high-leverage refresh candidates where traffic recovery is highly valuable and realistic.
3. **Age is Not a Simple Proxy**: The median age of declining pages is **216 days**, whereas non-declining pages have a median age of **287 days**. This counterintuitive result demonstrates that older content is not automatically the most prone to decay. A simple age-based cutoff rule would miss key declining assets and waste budget on stable older pages.

In [3]:
# Load starter data and extract key metrics
import pandas as pd
df = pd.read_csv("../../data/raw/content_refresh_anonymized.csv")

total_pages = len(df)
declining_mask = df["trend_direction"].str.lower().eq("down")
declining_pages = declining_mask.sum()
high_vis_declining = ((declining_mask) & (df["impressions_90d"] >= 500)).sum()
median_age_declining = df[declining_mask]["content_age_days"].median()
median_age_stable = df[~declining_mask]["content_age_days"].median()

print(f"1. Total Pages: {total_pages:,}")
print(f"2. Declining Pages: {declining_pages:,} ({declining_pages/total_pages:.1%})")
print(f"3. High-Visibility Declining Pages (impressions >= 500): {high_vis_declining:,} ({high_vis_declining/declining_pages:.1%} of declining)")
print(f"4. Median Age (Declining): {median_age_declining:.1f} days")
print(f"5. Median Age (Stable/Up): {median_age_stable:.1f} days")


1. Total Pages: 30,000
2. Declining Pages: 16,262 (54.2%)
3. High-Visibility Declining Pages (impressions >= 500): 9,961 (61.3% of declining)
4. Median Age (Declining): 216.0 days
5. Median Age (Stable/Up): 287.0 days


## 4. Careful words: what I can and can't claim

- **What I CAN claim**:
  - We can claim **observed associations** between historical search/engagement features (e.g., impressions, CTR, position, age) and page decline.
  - We can claim **decision-support** utility, helping editors prioritize their limited editing capacity by ranking pages by their historical probability of decline.
  - We can claim **directional** findings showing where traffic and ranking indicators suggest decay.
- **What I CANNOT claim**:
  - We cannot claim **causal proof**. We cannot state that updating a page *causes* a recovery, as that would require a randomized control experiment or causal inference modeling.
  - We cannot claim to have "predicted Google's algorithm" or identified official search engine ranking factors. Our analysis is strictly descriptive and client-specific search console telemetry.

In [4]:
# Demonstrate limit of simple linear correlations
import pandas as pd
df = pd.read_csv("../../data/raw/content_refresh_anonymized.csv")
y = df["trend_direction"].str.lower().eq("down").astype(int)

print("Correlation with decline label:")
print(f" - content_age_days: {df['content_age_days'].corr(y):.4f}")
print(f" - ctr: {df['ctr'].corr(y):.4f}")
print("The weak correlations indicate that a simple linear claim about single variables is insufficient.")


Correlation with decline label:
 - content_age_days: -0.1639
 - ctr: -0.0619
The weak correlations indicate that a simple linear claim about single variables is insufficient.


## Self-check

Before you submit, confirm each line honestly:

- [x] Every section above is filled — markdown thinking AND the code that backs it
- [x] The notebook runs top to bottom with no errors (Runtime → Run all)
- [x] No client names, URLs, or private queries anywhere
- [x] My claims use careful words: observed, measured, directional, decision-support
- [x] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.